### PR TITLE
Count publishers and subscribers

### PR DESCRIPTION
**Pull request by Nickolaim**
This is a partial implementation of [https://github.com/ros2/ros2cli/issues/60](https://github.com/ros2/ros2cli/issues/60).  I marked it as a WIP since I expect some comments.  Plan to merge changes later and do another PR.

Connects to ros2/ros2cli#60

### CODE REVIEW

**Comment by wjwwood** [comment]
Is this still a WIP? Let us know when this and related pr's are ready for review.

**Comment by wjwwood** [comment]
Sorry for closing, misclick on my part.

File: rclpy/rclpy/node.py [hunk 1]
```diff
@@ -295,5 +295,16 @@ def get_service_names_and_types(self):
     def get_node_names(self):
         return _rclpy.rclpy_get_node_names(self.handle)
 
+    def _count_publishers_or_subscribers(self, topic_name, func):
+        fq_topic_name = expand_topic_name(topic_name, self.get_name(), self.get_namespace())
+        validate_topic_name(fq_topic_name)
+        return func(self.handle, fq_topic_name)
+
+    def count_publishers(self, topic_name):
```
Comment (sloretz) [inline rclpy/rclpy/node.py]: Mind adding a doc string to this and `count_subscribers`? I realize most functions here don't have one, but adding them to new functions is an improvement.

Comment (sloretz) [reply rclpy/rclpy/node.py]: 66bdf9c

File: rclpy/test/test_node.py [hunk 2]
```diff
@@ -89,6 +92,33 @@ def test_node_names(self):
...
+        with self.assertRaisesRegex(TypeError, 'Argument topic_name is not a'):
+            self.node.count_subscribers(1)
+        with self.assertRaisesRegex(ValueError, 'is invalid'):
+            self.node.count_subscribers('42')
+        with self.assertRaisesRegex(ValueError, 'is invalid'):
+            self.node.count_publishers('42')
+
     def test_node_logger(self):
         node_logger = self.node.get_logger()
         self.assertEqual(node_logger.name, 'my_ns.my_node')
```
[the remark this answers is not in the document]
Comment (sloretz) [reply rclpy/test/test_node.py]: ca19854

**Comment by sloretz** [comment]
CI

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/4389/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1342/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3585/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4439/)

**Comment by sloretz** [comment]
@Nickolaim Would you mind rebasing your branch with master? CI looks ok except for test failures caused by not having #182.

**Comment by Nickolaim** [comment]
Sure, will do later today.  I am relatively new to github; for the request I plan to add remote upstream that points to ros2/rclpy and merge the changes from it.  Does it sound good?

**Comment by Karsten1987** [comment]
I assume here your remote is called `upstream`, points to `ros2/rclpy` and your current branch is `pubsub_count3`. Then the way to rebase is as follows.

```
# making sure you update all remotes
git fetch --all 
# rebase your changes on top of current master
git rebase upstream/master
```

**Comment by sloretz** [comment]
@Nickolaim adding on to @Karsten1987's answer, assuming your fork is called `origin` make sure to 

```
git checkout pubsub_count3
git pull origin pubsub_count3
```

then after rebasing

```
git push --force-with-lease origin pubsub_count3
```
```

**Comment by Nickolaim** [comment]
I rebased this PR with the master.

**Comment by sloretz** [comment]
CI
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/4398/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1350/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3594/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4455/)

**Comment by Nickolaim** [comment]
Looks like all the builds have passed.

**Pull request merged**